File: src/main/java/com/example/jwttest/jwt/JwtAccessDeniedHandler.java
package com.example.jwttest.jwt;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;


/**
 * "인가" 예외 처리. (403 Forbidden)
 * 권한이 맞지 않을 때 처리
 */
@Component
public class JwtAccessDeniedHandler implements AccessDeniedHandler {

    private static final Logger log = LoggerFactory.getLogger(JwtAuthenticationEntryPoint.class);

    @Override
    public void handle(HttpServletRequest request, HttpServletResponse response, AccessDeniedException accessDeniedException) throws IOException {
        response.setStatus(HttpServletResponse.SC_FORBIDDEN);
        response.setContentType("application/json");
        response.getWriter().write(String.format("{\"error\": \"%s\"}", "권한이 없습니다."));
        log.error("인가 거부. 권한이 올바르지 않음!!");
    }
}

File: src/main/kotlin/com/example/jwttest/jwt/JwtAccessDeniedHandler.kt
package com.example.jwttest.jwt

import org.slf4j.LoggerFactory
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component
import java.io.IOException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


/**
 * "인가" 예외 처리. (403 Forbidden)
 * 권한이 맞지 않을 때 처리
 */
@Component
class JwtAccessDeniedHandler : AccessDeniedHandler {

    private val log = LoggerFactory.getLogger(javaClass)
    @Throws(IOException::class)
    override fun handle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        accessDeniedException: AccessDeniedException
    ) {
        response.status = HttpServletResponse.SC_FORBIDDEN;
        response.contentType = "application/json";
        response.writer.write(String.format("{\"error\": \"%s\"}", "권한이 없습니다."))
        log.error("인가 거부. 권한이 올바르지 않음!!");
    }
}